### PR TITLE
[graphql] Allow for shorthand representation of the coin type

### PR DESCRIPTION
**Pull request by stefan-mysten**
## Description 

Allow for shorthand representation of the coin type.

## Test Plan 

image

On Error:
image
---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Review by wlmyng — APPROVED** [review]
nice

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs [hunk 1]
```diff
@@ -233,7 +233,9 @@ impl PgManager {
             .or(before)
             .map(|cursor| self.parse_obj_cursor(&cursor))
             .transpose()?;
-
+        let coin_type = parse_to_type_tag(Some(coin_type))
+            .map_err(|_| Error::Internal("Invalid coin type.".to_string()))?
```
Comment (oxade) [inline crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:237]: Can you propagate the actual error through?

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs [hunk 2]
```diff
@@ -861,13 +863,8 @@ impl PgManager {
...
         let address = address.into_vec();
-        let coin_type_tag = parse_to_type_tag(coin_type);
-        if coin_type_tag.is_err() {
-            // The provided `coin_type` cannot be parsed to a type tag so return None here.
-            return Ok(None);
-        }
-        let coin_type = coin_type_tag
-            .unwrap()
+        let coin_type = parse_to_type_tag(coin_type)
+            .map_err(|_| Error::Internal("Invalid coin type.".to_string()))?
```
Comment (oxade) [inline crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:867]: Same here on error propagation

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs (hunk 1 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:237]: Ah wait yeah, this should typically be a client error, +1

**Comment by stefan-mysten** [comment]
Thanks! 
Here's the output with error propagation. 
image

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs [hunk 3]
```diff
@@ -233,7 +233,9 @@ impl PgManager {
             .or(before)
             .map(|cursor| self.parse_obj_cursor(&cursor))
             .transpose()?;
-
+        let coin_type = parse_to_type_tag(Some(coin_type))
+            .map_err(|e| Error::Internal(e.to_string()))?
```
Comment (wlmyng) [inline crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:237]: can you use Error::InvalidCoinType

**Review by wlmyng — CHANGES_REQUESTED** [review]

**Comment by wlmyng** [comment]
a caller providing 0x2::sui shouldn't be considered an internal error, the language misrepresents who is at fault for the error response

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs (hunk 3 above)
Comment (amnn) [reply crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:237]: Note that in some cases, this is an internal error, and in some cases this is a user error -- it depends on the callee, this is why the DB layer needs its own error type.

Comment (amnn) [reply crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:237]: Until it does, I'm OK with us using either error type here -- neither is ideal, unfortunately.

**Comment by stefan-mysten** [comment]
> a caller providing 0x2::sui shouldn't be considered an internal error, the language misrepresents who is at fault for the error response

Right, so then I'd add a new error to our error type.

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs (hunk 3 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:237]: ok that's fair
we have a DbValidationError enum that we could add InvalidCoinType to

Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:237]: image

**Review by wlmyng — APPROVED** [review]
like @amnn in the query layer its hard to know whether the fault is with client or if the object read has an invalid coin type, so representing it either as internal or client error is fine. Although today parsing a coin type is predominantly done on a user input

**Comment by stefan-mysten** [comment]
I agree with both of you. We'll have to revisit this later when we move this conversion to the reader side.

**Comment by wlmyng** [comment]
> Note that in some cases, this is an internal error, and in some cases this is a user error -- it depends on the caller, this is why the DB layer needs its own error type.

I think it makes sense for db_data_provider to control whether this is surfaced as an internal or client error, mainly because it sits between graphql and the actual db reader (IndexerReader). When the InvalidCoinType occurs on an arg provided to a db_data_provider method, it'll always return the error as a client error. If the parsing error occurs on an object read from the table, then it will be treated as an internal error. Then,
1. If this is used by another db_data_provider method, that method also has enough context to determine whether it is a user error or not. Say method_b fetched an object and passed its coin_type to db_data_provider.method_a, method_a returns client error but method_b knows its an internal error.
3. The layer above the db_data_provider would be the graphql resolver fields, which may not have all the context needed to determine error kind. If it sees an InvalidCoinType, it may not know if its due to the args given to itself (the resolver) or if it came from somewhere else down the line

**Pull request merged**